File: java/src/main/java/toolgood/flowVision/Engines/OutDatas/TreeNode.java
package toolgood.flowVision.Engines.OutDatas;

import toolgood.algorithm2.Operand;
import toolgood.flowVision.Engines.FlowEngine;
import toolgood.flowVision.Flows.Enums.CellType;
import toolgood.flowVision.Flows.Enums.InputNumberType;
import toolgood.flowVision.Flows.FactoryMachineWork;
import toolgood.flowVision.Flows.FactoryProcedureItemWork;
import toolgood.flowVision.Flows.Interfaces.IInputFormulaNodeWork;
import toolgood.flowVision.Flows.NodeWork;
import toolgood.flowVision.Flows.ProcedureFlowWork;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;


public class TreeNode {
    public List<ChannelNode> PrevNodes;
    public Map<String, TreeNode> NextNodes;
    public NodeWork CurrWork;
    public FactoryMachineWork MachineItem;
    public FactoryProcedureItemWork ProcedureItem;
    public BigDecimal InputNum;
    public BigDecimal OutputNum;

    public TreeNode(NodeWork work, FactoryMachineWork factoryMachineItem, FactoryProcedureItemWork factoryProcedureItem) {
        CurrWork = work;
        MachineItem = factoryMachineItem;
        ProcedureItem = factoryProcedureItem;
        PrevNodes = new ArrayList<>();
        NextNodes = new HashMap<>();
    }

    public String Id() {
        return CurrWork.Id;
    }

    public int Layer() {
        return CurrWork.Layer;
    }

    public boolean IsSubsidiaryCount() {
        if (CurrWork instanceof IInputFormulaNodeWork) {
            IInputFormulaNodeWork work = (IInputFormulaNodeWork) CurrWork;
            return work.IsSubsidiaryCount();
        }
        return false;
    }

    public void AddNextNode(TreeNode node, String channel, int index) {
        NextNodes.put(channel, node);
        ChannelNode nd = new ChannelNode();
        nd.Index = index;
        nd.Node = node;
        nd.Channel = channel;
        node.PrevNodes.add(nd);
    }

    public void SetError(String channel) {
        for (int i = 0; i < PrevNodes.size(); i++) {
            ChannelNode prevNode = PrevNodes.get(i);
            prevNode.Node.NextNodes.remove(channel);
        }
    }

    public void EvaluateInputNum(FlowEngine engine) throws Exception {
        if (InputNum != null) {
            return;
        }
        if (CurrWork.NodeType == CellType.End) {
            OutputNum = engine.GetNum().NumberValue();
            if (CurrWork instanceof ProcedureFlowWork) {
                ProcedureFlowWork procedure = (ProcedureFlowWork) CurrWork;
                OutputNum =new BigDecimal( GetNumber(procedure.NumberType, OutputNum.doubleValue()));
            }
            InputNum = OutputNum;
            return;
        }
        if (OutputNum == null) {
            BigDecimal num = new BigDecimal(0);
            for (TreeNode nextNode : NextNodes.values()) {
                nextNode.EvaluateInputNum(engine);
                if (nextNode.IsSubsidiaryCount() == false) {
                    if (num.compareTo(nextNode.InputNum) ==-1) {
                        num = nextNode.InputNum;
                    }
                }
            }
            OutputNum = num;
        }
        engine.SetOutputNum(Operand.Create(OutputNum));
        if (InputNum == null) {
            if (CurrWork instanceof IInputFormulaNodeWork) {
                IInputFormulaNodeWork work = (IInputFormulaNodeWork) CurrWork;
                InputNum = work.EvaluateInputNum(engine);
            } else {
                InputNum = OutputNum;
            }
            if (CurrWork instanceof ProcedureFlowWork) {
                ProcedureFlowWork procedure = (ProcedureFlowWork) CurrWork;
                InputNum =new BigDecimal(GetNumber(procedure.NumberType, InputNum.doubleValue())) ;
            }
        }
        engine.ClearOutputNum();
    }

    private double GetNumber(InputNumberType numberType, double n) {
        switch (numberType) {
            case Original:
                return n;
            case Ceil:
                return Math.ceil(n);
            case Floor:
                return Math.floor(n);
            case Round0:
                return Math.round(n);
            case Round1:
                return Math.round(n * 10) / 10d;
            case Round2:
                return Math.round(n * 100) / 100d;
            case Round3:
                return Math.round(n * 1000) / 1000d;
            case Round4:
                return Math.round(n * 10000) / 10000d;
            case Round5:
                return Math.round(n * 100000) / 100000d;
            case Round6:
                return Math.round(n * 1000000) / 1000000d;
            case Round7:
                return Math.round(n * 10000000) / 10000000d;
            case Round8:
                return Math.round(n * 100000000) / 100000000d;
            default:
                return n;
        }
    }

}

File: editor/ToolGood.FlowWork.Applications/Engines/OutDatas/TreeNode.cs
using ToolGood.FlowWork.Flows;

namespace ToolGood.FlowWork.Applications.Engines.OutDatas
{
	public sealed class TreeNode
	{
		public List<ChannelNode> PrevNodes { get; private set; }
		public Dictionary<string, TreeNode> NextNodes { get; private set; }

		public NodeWork CurrWork { get; private set; }
		public FactoryMachineWork MachineItem { get; private set; }
		public FactoryProcedureItemWork ProcedureItem { get; private set; }

		public string Id { get { return CurrWork.Id; } }
		public int Layer { get { return CurrWork.Layer; } }
		public bool IsSubsidiaryCount { get { if (CurrWork is IInputFormulaNodeWork work) { return work.IsSubsidiaryCount; } return false; } }
		public decimal? InputNum { get; set; }
		public decimal? OutputNum { get; set; }

		public TreeNode(NodeWork work, FactoryMachineWork factoryMachineItem, FactoryProcedureItemWork factoryProcedureItem)
		{
			CurrWork = work;
			MachineItem = factoryMachineItem;
			ProcedureItem = factoryProcedureItem;
			PrevNodes = new List<ChannelNode>();
			NextNodes = new Dictionary<string, TreeNode>();
		}

		public void AddNextNode(TreeNode node, string channel, int index)
		{
			NextNodes.Add(channel, node);
			node.PrevNodes.Add(new ChannelNode {
				Channel = channel,
				Index = index,
				Node = this
			});
		}

		public void SetError(string channel)
		{
			for (int i = 0; i < PrevNodes.Count; i++) {
				var prevNode = PrevNodes[i];
				prevNode.Node.NextNodes.Remove(channel);
			}
		}

		#region 计算入量

		public void EvaluateInputNum(FlowEngine engine)
		{
			if (InputNum != null) { return; }
			if (CurrWork.NodeType == CellType.End) {
				OutputNum = (decimal)engine.GetNum().NumberValue;
				if (CurrWork is ProcedureFlowWork procedure) {
					OutputNum = GetNumber(procedure.NumberType, OutputNum.Value);
				}
				InputNum = OutputNum;
				return;
			}
			if (OutputNum == null) {
				decimal num = 0;
				foreach (var nextNode in NextNodes) {
					nextNode.Value.EvaluateInputNum(engine);
					if (nextNode.Value.IsSubsidiaryCount == false) {
						if (num < nextNode.Value.InputNum.Value) {
							num = nextNode.Value.InputNum.Value;
						}
					}
				}
				OutputNum = num;
			}
			engine.SetOutputNum(OutputNum.Value);
			if (InputNum == null) {
				if (CurrWork is IInputFormulaNodeWork work) {
					InputNum = work.EvaluateInputNum(engine);
				} else {
					InputNum = OutputNum;
				}
				if (CurrWork is ProcedureFlowWork procedure) {
					InputNum = GetNumber(procedure.NumberType, InputNum.Value);
				}
			}
			engine.ClearOutputNum();
		}

		private decimal GetNumber(InputNumberType numberType, decimal n)
		{
			switch (numberType) {
				case InputNumberType.Original: return n;
				case InputNumberType.Ceil: return Math.Ceiling(n);
				case InputNumberType.Floor: return Math.Floor(n);
				case InputNumberType.Round0: return Math.Round(n, 0, MidpointRounding.AwayFromZero);
				case InputNumberType.Round1: return Math.Round(n, 1, MidpointRounding.AwayFromZero);
				case InputNumberType.Round2: return Math.Round(n, 2, MidpointRounding.AwayFromZero);
				case InputNumberType.Round3: return Math.Round(n, 3, MidpointRounding.AwayFromZero);
				case InputNumberType.Round4: return Math.Round(n, 4, MidpointRounding.AwayFromZero);
				case InputNumberType.Round5: return Math.Round(n, 5, MidpointRounding.AwayFromZero);
				case InputNumberType.Round6: return Math.Round(n, 6, MidpointRounding.AwayFromZero);
				case InputNumberType.Round7: return Math.Round(n, 7, MidpointRounding.AwayFromZero);
				case InputNumberType.Round8: return Math.Round(n, 8, MidpointRounding.AwayFromZero);
				default: return n;
			}
		}

		#endregion 计算入量
	}

	public sealed class StockNode
	{
		private List<ChannelNode> Nodes = new List<ChannelNode>();
		private HashSet<string> IdSet = new HashSet<string>();

		public void Push(TreeNode node, string channel, int index)
		{
			if (IdSet.Add(node.CurrWork.Id)) {// 防止合并
				Nodes.Add(new ChannelNode {
					Channel = channel,
					Index = index,
					Node = node
				});
			}
		}

		public bool TryPop(out ChannelNode node)
		{
			if (Nodes.Count == 0) { node = null; return false; }

			ChannelNode temp = Nodes[0];
			for (int i = 1; i < Nodes.Count; i++) {
				var channelNode = Nodes[i];
				if (temp != null) {
					if (channelNode.Node.Layer < temp.Node.Layer) {
						temp = channelNode;
					}
				} else {
					temp = channelNode;
				}
			}
			node = temp;
			if (temp != null) {
				Nodes.Remove(temp);
				IdSet.Remove(temp.Node.Id);
				return true;
			}
			return false;
		}
	}

	/// <summary>
	/// Channel，Index 为下个节点所在位置
	/// 如查将 Channel，Index 放在 TreeNode ，多条工艺渠道合并就会出错
	/// </summary>
	public sealed class ChannelNode
	{
		public TreeNode Node;
		public string Channel;
		public int Index;
	}
}